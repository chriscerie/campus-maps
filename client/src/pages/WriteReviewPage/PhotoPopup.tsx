import Modal from '@mui/material/Modal';
import { useState } from 'react';
import CloseButton from '../../components/CloseButton';
import './PhotoPopup.scss';

function PhotoPopup({
  photos,
  index,
  setIndex,
  onClose,
}: {
  photos: Array<{ imageSrc: string; file: File }>;
  index: number;
  setIndex: React.Dispatch<React.SetStateAction<number>>;
  onClose: (event: {}) => void;
}) {
  const [imageMousePos, setImageMousePos] = useState<'Left' | 'Right'>('Right');

  return (
    <Modal
      open={index >= 0}
      onClose={onClose}
      sx={{
        padding: '3em 1em',
        '@media (min-width: 769px)': {
          padding: '3em 2em',
        },
      }}
    >
      <div id="photo-popup-container">
        <div id="photo-popup-inner-container">
          <CloseButton
            sx={{
              position: 'absolute',
              right: '1em',
              top: '1em',
            }}
            onClick={() => {
              onClose({});
            }}
          />
          <div id="photo-popup-inner">
            <div
              id="photo-popup-gallery"
              onMouseMove={(e) => {
                const posX =
                  e.clientX - e.currentTarget.getBoundingClientRect().left;
                const isLeftSide = posX < e.currentTarget.clientWidth / 2;
                setImageMousePos(isLeftSide ? 'Left' : 'Right');
              }}
              onClick={() => {
                imageMousePos === 'Left'
                  ? setIndex((prev) => Math.max(0, prev - 1))
                  : setIndex((prev) => Math.min(photos.length - 1, prev + 1));
              }}
            >
              <div id="photo-popup-photo-background">
                <div id="text-area">
                  <pre style={{ margin: 0, userSelect: 'none' }}>
                    {imageMousePos === 'Left' &&
                      `<  ${index + 1} of ${photos.length}   `}
                    {imageMousePos === 'Right' &&
                      `   ${index + 1} of ${photos.length}  >`}
                  </pre>
                </div>
              </div>
              <img
                id="photo-popup-photo"
                alt="Location"
                src={photos[index] && photos[index].imageSrc}
              />
            </div>
            <div id="photo-popup-sidebar-container"></div>
          </div>
        </div>
      </div>
    </Modal>
  );
}

export default PhotoPopup;
