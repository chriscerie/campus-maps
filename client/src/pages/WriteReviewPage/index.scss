#write-review-page-container {
  display: table;
  margin-top: 4em;
  margin-bottom: 10em;
}

#write-review-page-inner {
  float: left;
  box-sizing: border-box;
  width: 100%;
  max-width: 550px;
  text-align: left;
}

%write-review-textinput {
  display: block;
  margin: 0 0 1em;
  padding: 0.4em 0.8em;
  border: 1px solid #999;
  border-radius: 0.3em;
  width: 100%;
  font-size: 1em;
  box-sizing: border-box;
}

%write-review-textinput:focus {
  outline: none;
}

#write-review-page-inner p {
  margin-bottom: 2em;
}

#write-review-page-inner textarea {
  @extend %write-review-textinput;
  resize: vertical;
  height: 20em;
  padding: 1.5em;
}

#write-review-submit-button {
  color: #fff;
  background-color: #1a8ceb;
  border: 0;
  padding: 0.4em 1.4em;
  border-radius: 0.3em;
  font-weight: bold;
  font-size: 1em;
  text-transform: none;
}

#write-review-submit-button:hover {
  background-color: #2592ff;
}

#write-review-cancel-link {
  color: #1577da;
  text-decoration: none;
  font-size: 1em;
  margin-left: 1em;
}

#write-review-cancel-link:hover {
  text-decoration: underline;
}
