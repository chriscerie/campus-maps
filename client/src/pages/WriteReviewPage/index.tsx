import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../reducers';
import { useParams, Link, Redirect } from 'react-router-dom';
import React, { useState, useEffect } from 'react';
import { Container, Button } from '@mui/material';
import UploadFileIcon from '@mui/icons-material/UploadFile';
import DeleteIcon from '@mui/icons-material/Delete';
import { useForm } from 'react-hook-form';
import axios from 'axios';
import { getLocationInfo } from '../../api/LocationAPI';
import { getBase64, getFilesFromBase64 } from '../../api/GetBase64';
import { setCurrentUser } from '../../actions/currentUserActions';
import type { LocationType } from '../../types/LocationType';
import type { ReviewType } from '../../types/ReviewType';
import PhotoPopup from '../../components/PhotoPopup';
import './index.scss';

function WriteReviewPage() {
  const { id } = useParams<{ id: string }>();
  const [locationInfo, setLocationInfo] = useState<null | LocationType>(null);
  const mapInstance = useSelector((state: RootState) => state.mapInstance.map);
  const currentUser = useSelector((state: RootState) => state.currentUser);
  const { register, setValue, handleSubmit } = useForm<{
    body: string;
  }>();
  const [photos, setPhotos] = useState<Array<{ imageSrc: string; file: File }>>(
    []
  );
  const [submitted, setSubmitted] = useState(false);
  const [selectedPhotoIndex, setSelectedPhotoIndex] = useState<number>(-1);
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(setCurrentUser());
  }, [dispatch]);

  // Set review
  useEffect(() => {
    if (locationInfo && currentUser) {
      axios
        .get<ReviewType>(`/api/v1/reviews/${id}`, {
          params: { filter_by: 'current_user' },
        })
        .then(async (res) => {
          setValue('body', res.data.body);
        });

      // Get all currentUser.photos that match location_id with id
      const currentPhotos = currentUser.photos.filter(
        (photo) => photo.location_id === id
      );

      // Set photos
      getFilesFromBase64(
        currentPhotos.map((photo) => {
          return photo.photo;
        })
      ).then((photos) => {
        setPhotos(photos);
      });
    }
  }, [id, locationInfo, currentUser, setValue]);

  // Set point
  useEffect(() => {
    if (mapInstance) {
      getLocationInfo(id, mapInstance).then((res) => {
        if (res) {
          setLocationInfo(res);
        }
      });
    }
  }, [id, mapInstance]);

  const onSubmit = (e: React.FormEvent) => {
    e.preventDefault();

    handleSubmit((data) => {
      const imageBase64s: Array<string> = [];

      Promise.all(
        photos.map(async (photo) => {
          const base64 = await getBase64(photo.file);
          imageBase64s.push(base64);
        })
      ).then(() => {
        axios
          .post(`/api/v1/review/${id}`, {
            body: data.body,
            photos: imageBase64s,
          })
          .then((res) => {
            setSubmitted(true);
          })
          .catch((err) => {
            console.log(err);
          });
      });
    })();
  };

  return submitted ? (
    <Redirect to={`/loc/${id}`} />
  ) : (
    <Container maxWidth="md" id="write-review-page-container">
      <div id="write-review-page-inner">
        <Link
          to={`/loc/${id}`}
          style={{ textDecoration: 'none', color: '#000' }}
        >
          <h1 style={{ fontSize: '2em' }}>
            {locationInfo && locationInfo.name}
          </h1>
        </Link>
        <p>All reviews are subject to moderation.</p>
        <form onSubmit={onSubmit}>
          <ul>
            <li key="Review" style={{ marginBottom: '3em' }}>
              <textarea {...register('body')} placeholder="Write a review." />
            </li>
            <li key="Photos">
              <h3 style={{ fontSize: '1.3em' }}>
                Attach Photos (32 KB file limit)
              </h3>
              <div id="write-review-photos-section">
                <div style={photos.length === 0 ? { width: '100%' } : {}}>
                  <div
                    id="write-review-add-photos-container"
                    style={photos.length > 0 ? { width: '8.2em' } : {}}
                  >
                    <UploadFileIcon
                      sx={{
                        margin: '1.5em 0',
                        width: '1.2em',
                        height: '1.2em',
                      }}
                    />
                    <input
                      type="file"
                      multiple
                      accept=".png, .jpg, .jpeg"
                      id="write-review-add-photos"
                      onChange={(e) => {
                        if (e.target.files) {
                          // Only allow up to 32 KB
                          const files = Array.from(e.target.files).filter(
                            (file) => file.size < 32000
                          );

                          // Alert if any file got filtered
                          if (
                            files.length < Array.from(e.target.files).length
                          ) {
                            alert(
                              'File size is too large. File limit is 32 KB.'
                            );
                          }

                          setPhotos((prev) =>
                            prev.concat(
                              files.map((file) => ({
                                imageSrc: URL.createObjectURL(file),
                                file: file,
                              }))
                            )
                          );
                        }
                      }}
                    />
                  </div>
                </div>
                <PhotoPopup
                  photos={photos}
                  index={selectedPhotoIndex}
                  setIndex={setSelectedPhotoIndex}
                  onClose={(e) => {
                    setSelectedPhotoIndex(-1);
                  }}
                />
                {photos &&
                  photos.map((photo, index) => (
                    <div
                      className="write-review-photo-container"
                      style={{
                        backgroundImage: `url(${photo.imageSrc})`,
                        width: '8.2em',
                        height: '8.2em',
                      }}
                    >
                      <div className="write-review-photo-background">
                        <div
                          className="text-area"
                          onClick={() => {
                            setSelectedPhotoIndex(index);
                          }}
                        >
                          <p style={{ margin: 0, fontWeight: 700 }}>
                            Add caption
                          </p>
                        </div>
                      </div>
                      <div
                        className="write-review-photo-delete-icon"
                        onClick={() => {
                          setPhotos((prev) => [
                            ...prev.slice(0, index),
                            ...prev.slice(index + 1),
                          ]);
                        }}
                      >
                        <DeleteIcon />
                      </div>
                    </div>
                  ))}
              </div>
            </li>
          </ul>
          <div style={{ marginTop: '2em' }}>
            <Button
              variant="contained"
              disableElevation
              id="write-review-submit-button"
              type="submit"
            >
              Post Review
            </Button>
            <Link to={`/loc/${id}`} id="write-review-cancel-link">
              Cancel
            </Link>
          </div>
        </form>
      </div>
    </Container>
  );
}

export default WriteReviewPage;
