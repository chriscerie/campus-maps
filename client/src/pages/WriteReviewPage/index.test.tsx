import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import store from '../../reducers';
import WriteReviewPage from './index';

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as object),
  useParams: jest.fn().mockReturnValue({ id: '1111' }),
}));

describe('WriteReviewPage', () => {
  it('should render successfully', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <WriteReviewPage />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
