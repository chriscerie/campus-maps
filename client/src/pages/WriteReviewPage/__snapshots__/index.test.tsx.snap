// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WriteReviewPage should render successfully 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-xdmu94-MuiContainer-root{width:100%;margin-left:auto;box-sizing:border-box;margin-right:auto;display:block;padding-left:16px;padding-right:16px;}@media (min-width:600px){.css-xdmu94-MuiContainer-root{padding-left:24px;padding-right:24px;}}@media (min-width:900px){.css-xdmu94-MuiContainer-root{max-width:900px;}}",
      }
    }
    data-emotion="css xdmu94-MuiContainer-root"
  />,
  <div
    className="MuiContainer-root MuiContainer-maxWidthMd css-xdmu94-MuiContainer-root"
    id="write-review-page-container"
  >
    <div
      id="write-review-page-inner"
    >
      <a
        href="/loc/1111"
        onClick={[Function]}
        style={
          Object {
            "color": "#000",
            "textDecoration": "none",
          }
        }
      >
        <h1
          style={
            Object {
              "fontSize": "2em",
            }
          }
        />
      </a>
      <p>
        All reviews are subject to moderation.
      </p>
      <form>
        <ul>
          <li
            style={
              Object {
                "marginBottom": "3em",
              }
            }
          >
            <textarea
              name="body"
              placeholder="Write a review."
            />
          </li>
          <li>
            <h3
              style={
                Object {
                  "fontSize": "1.3em",
                }
              }
            >
              Attach Photos
            </h3>
            <div
              id="write-review-photos-section"
            >
              <div
                style={
                  Object {
                    "width": "100%",
                  }
                }
              >
                <div
                  id="write-review-add-photos-container"
                  style={Object {}}
                >
                  <style
                    dangerouslySetInnerHTML={
                      Object {
                        "__html": ".css-18xo0gl-MuiSvgIcon-root{-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;width:1em;height:1em;display:inline-block;fill:currentColor;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;-webkit-transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;font-size:1.5rem;margin:1.5em 0;width:1.2em;height:1.2em;}",
                      }
                    }
                    data-emotion="css 18xo0gl-MuiSvgIcon-root"
                  />
                  <svg
                    aria-hidden={true}
                    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-18xo0gl-MuiSvgIcon-root"
                    data-testid="UploadFileIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm4 18H6V4h7v5h5v11zM8 15.01l1.41 1.41L11 14.84V19h2v-4.16l1.59 1.59L16 15.01 12.01 11z"
                    />
                  </svg>
                  <input
                    accept=".png, .jpg, .jpeg"
                    id="write-review-add-photos"
                    multiple={true}
                    onChange={[Function]}
                    type="file"
                  />
                </div>
              </div>
            </div>
          </li>
        </ul>
        <div
          style={
            Object {
              "marginTop": "2em",
            }
          }
        >
          <style
            dangerouslySetInnerHTML={
              Object {
                "__html": ".css-1diear7-MuiButton-root{font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:500;font-size:0.875rem;line-height:1.75;letter-spacing:0.02857em;text-transform:uppercase;min-width:64px;padding:6px 16px;border-radius:4px;-webkit-transition:background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;color:#fff;background-color:#1976d2;box-shadow:0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);box-shadow:none;}.css-1diear7-MuiButton-root:hover{-webkit-text-decoration:none;text-decoration:none;background-color:#1565c0;box-shadow:0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);}@media (hover: none){.css-1diear7-MuiButton-root:hover{background-color:#1976d2;}}.css-1diear7-MuiButton-root:active{box-shadow:0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);}.css-1diear7-MuiButton-root.Mui-focusVisible{box-shadow:0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);}.css-1diear7-MuiButton-root.Mui-disabled{color:rgba(0, 0, 0, 0.26);box-shadow:none;background-color:rgba(0, 0, 0, 0.12);}.css-1diear7-MuiButton-root:hover{box-shadow:none;}.css-1diear7-MuiButton-root.Mui-focusVisible{box-shadow:none;}.css-1diear7-MuiButton-root:active{box-shadow:none;}.css-1diear7-MuiButton-root.Mui-disabled{box-shadow:none;}",
              }
            }
            data-emotion="css 1diear7-MuiButton-root"
          />
          <style
            dangerouslySetInnerHTML={
              Object {
                "__html": ".css-zddlty-MuiButtonBase-root-MuiButton-root{display:-webkit-inline-box;display:-webkit-inline-flex;display:-ms-inline-flexbox;display:inline-flex;-webkit-align-items:center;-webkit-box-align:center;-ms-flex-align:center;align-items:center;-webkit-box-pack:center;-ms-flex-pack:center;-webkit-justify-content:center;justify-content:center;position:relative;box-sizing:border-box;-webkit-tap-highlight-color:transparent;background-color:transparent;outline:0;border:0;margin:0;border-radius:0;padding:0;cursor:pointer;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;vertical-align:middle;-moz-appearance:none;-webkit-appearance:none;-webkit-text-decoration:none;text-decoration:none;color:inherit;font-family:\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif;font-weight:500;font-size:0.875rem;line-height:1.75;letter-spacing:0.02857em;text-transform:uppercase;min-width:64px;padding:6px 16px;border-radius:4px;-webkit-transition:background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;transition:background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;color:#fff;background-color:#1976d2;box-shadow:0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);box-shadow:none;}.css-zddlty-MuiButtonBase-root-MuiButton-root::-moz-focus-inner{border-style:none;}.css-zddlty-MuiButtonBase-root-MuiButton-root.Mui-disabled{pointer-events:none;cursor:default;}@media print{.css-zddlty-MuiButtonBase-root-MuiButton-root{-webkit-print-color-adjust:exact;color-adjust:exact;}}.css-zddlty-MuiButtonBase-root-MuiButton-root:hover{-webkit-text-decoration:none;text-decoration:none;background-color:#1565c0;box-shadow:0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);}@media (hover: none){.css-zddlty-MuiButtonBase-root-MuiButton-root:hover{background-color:#1976d2;}}.css-zddlty-MuiButtonBase-root-MuiButton-root:active{box-shadow:0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);}.css-zddlty-MuiButtonBase-root-MuiButton-root.Mui-focusVisible{box-shadow:0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);}.css-zddlty-MuiButtonBase-root-MuiButton-root.Mui-disabled{color:rgba(0, 0, 0, 0.26);box-shadow:none;background-color:rgba(0, 0, 0, 0.12);}.css-zddlty-MuiButtonBase-root-MuiButton-root:hover{box-shadow:none;}.css-zddlty-MuiButtonBase-root-MuiButton-root.Mui-focusVisible{box-shadow:none;}.css-zddlty-MuiButtonBase-root-MuiButton-root:active{box-shadow:none;}.css-zddlty-MuiButtonBase-root-MuiButton-root.Mui-disabled{box-shadow:none;}",
              }
            }
            data-emotion="css zddlty-MuiButtonBase-root-MuiButton-root"
          />
          <button
            className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-disableElevation MuiButtonBase-root css-zddlty-MuiButtonBase-root-MuiButton-root"
            disabled={false}
            id="write-review-submit-button"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            Post Review
          </button>
          <a
            href="/loc/1111"
            id="write-review-cancel-link"
            onClick={[Function]}
          >
            Cancel
          </a>
        </div>
      </form>
    </div>
  </div>,
]
`;
